raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-        ApiMessage responseData = response.data();
-        if (responseData instanceof FetchQuorumRecordsResponseData) {
-            handleFetchQuorumRecordsResponse((FetchQuorumRecordsResponseData) responseData);
-        } else if (responseData instanceof VoteResponseData) {
-            handleVoteResponse(response.sourceId(), (VoteResponseData) responseData);
-        } else if (responseData instanceof BeginQuorumEpochResponseData) {
-            handleBeginQuorumEpochResponse(response.sourceId());
-        } else if (responseData instanceof EndQuorumEpochResponseData) {
-            // Nothing to do for now
-        } else if (responseData instanceof FindQuorumResponseData) {
-            handleFindQuorumResponse((FindQuorumResponseData) responseData);
+        ApiKeys apiKey = ApiKeys.forId(response.data.apiKey());
+        final boolean handledSuccessfully;
+
+        switch (apiKey) {
+            case FETCH_QUORUM_RECORDS:
+                handledSuccessfully = handleFetchQuorumRecordsResponse(response);
+                break;
+
+            case VOTE:
+                handledSuccessfully = handleVoteResponse(response);
+                break;
+
+            case BEGIN_QUORUM_EPOCH:
+                handledSuccessfully = handleBeginQuorumEpochResponse(response);
+                break;
+
+            case END_QUORUM_EPOCH:
+                handledSuccessfully = handleEndQuorumEpochResponse(response);
+                break;
+
+            case FIND_QUORUM:
+                handledSuccessfully = handleFindQuorumResponse(response);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Received unexpected response type: " + apiKey);
+        }
+
+        ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (handledSuccessfully) {
+            connection.onResponseReceived(response.correlationId);
         } else {
-            throw new IllegalStateException("Received unexpected response " + response);
+            connection.onResponseError(response.correlationId, currentTimeMs);
         }
     }
 
-    private Optional<Exception> validateLeaderOnlyRequest(int remoteNodeId, int requestEpoch) throws IOException {
-        if (quorum.isObserver()) {
-            return Optional.of(Errors.INVALID_REQUEST.exception());
-        } else if (!quorum.isVoter(remoteNodeId)) {
-            return Optional.of(Errors.INVALID_REQUEST.exception());
-        } else if (requestEpoch < quorum.epoch()) {
-            return Optional.of(Errors.FENCED_LEADER_EPOCH.exception());
+    /**
+     * Validate a request which is only valid between voters.
+     */
+    private Optional<Errors> validateVoterOnlyRequest(int remoteNodeId, int requestEpoch) throws IOException {
+        if (requestEpoch < quorum.epoch()) {
+            return Optional.of(Errors.FENCED_LEADER_EPOCH);
+        } else if (quorum.isObserver() || !quorum.isVoter(remoteNodeId)) {
```

Comment:
I feel these two scenarios are not quite the same, as the former indicates a state error of the receiver, while the later indicates a sender's error. Does `UNKNOWN_VOTER` fit better for the latter case?

Reply:
Originally I did make this into separate errors, but I talked myself out of it because I could not think of a reason why the behavior would be different when the response is received. Either way, I think we're just going to handle this error by backing off and retrying.